### DISCUSSION TITLE
Remove Form XObject from PDF

### DISCUSSION THREAD

**Discussion by eddprogrammer** [pymupdf/PyMuPDF#2116] [Looking for help]
Hi, 
First I would like to thank you for such a great project, learning something new every day as I use it more!

I would like to remove some Form XObject from the PDF file and not sure how to do that. I can get the xref by calling page.get_xobjects() and inspect the object info with doc.xref_object method. How do I remove it from the PDF file? 

This is what I tried so far
`doc.xref_set_key(page.xref, "Resources/XObject", "null")`
And it got me exactly what I wanted, but I got an warning when opening the PDF in Adobe. 
<img width="420" alt="image" src="https://user-images.githubusercontent.com/119921381/206928291-f5a44e31-1a4c-4d89-b116-bc3af23edb21.png">

I can use `page.clean_contents()` before saving and it seems to get rid of that warning. 

Just not sure if that is the right way to handle. 

Thanks!

**Comment by JorjMcKie** [comment]
You are lucky that the problem got solved in this way. As it worked, your PDF should be okay.
for some background:
What you actually did remove, was the pointer to the xref of the XObject. Somewhere inside the possibly many `/Contents` objects, there still existed a reference to that XObject in the form of a command like `/refname Do`.
This is what caused that message. The `clean_contents()` method then cleaned up the situation.
If you instead would have manipulated the `/Contents` and removed that referencing instruction, there would have no message. But this always is a shaky thing to do, because of the possibly many `/Contents` which in addition may be arbitrarily nested.
So it was good luck it worked that way.

**Reply by eddprogrammer** [reply]
Appreciate the response there, hope there will be a more elegant way to remove both the reference in `/Content` and the the pointer to the xref of the XObject. I have seen in the other discussion on how to remove the referencing instruction, but if `clean_contents()` takes care of that, then that would work :)

Thanks again!

**Comment by occludedpixel** [comment]
Hi, is there any update on this? Specifically, I am trying to delete XObjects created using `Page.show_pdf_page()`.

**Reply by JorjMcKie** [reply]
No updates here.
